stackslib/src/net/api/getstackers.rs

```diff
@@ -0,0 +1,225 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+use regex::{Captures, Regex};
+use serde_json::json;
+use stacks_common::types::chainstate::StacksBlockId;
+use stacks_common::types::net::PeerHost;
+use stacks_common::util::hash::Sha256Sum;
+
+use crate::burnchains::Burnchain;
+use crate::chainstate::burn::db::sortdb::SortitionDB;
+use crate::chainstate::coordinator::OnChainRewardSetProvider;
+use crate::chainstate::stacks::boot::{
+    PoxVersions, RewardSet, POX_1_NAME, POX_2_NAME, POX_3_NAME, POX_4_NAME,
+};
+use crate::chainstate::stacks::db::StacksChainState;
+use crate::chainstate::stacks::Error as ChainError;
+use crate::core::mempool::MemPoolDB;
+use crate::net::http::{
+    parse_json, Error, HttpBadRequest, HttpNotFound, HttpRequest, HttpRequestContents,
+    HttpRequestPreamble, HttpResponse, HttpResponseContents, HttpResponsePayload,
+    HttpResponsePreamble, HttpServerError,
+};
+use crate::net::httpcore::{
+    HttpPreambleExtensions, HttpRequestContentsExtensions, RPCRequestHandler, StacksHttp,
+    StacksHttpRequest, StacksHttpResponse,
+};
+use crate::net::p2p::PeerNetwork;
+use crate::net::{Error as NetError, StacksNodeState, TipRequest};
+use crate::util_lib::boot::boot_code_id;
+use crate::util_lib::db::Error as DBError;
+
+#[derive(Clone, Default)]
+pub struct GetStackersRequestHandler {
+    cycle_number: Option<u64>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GetStackersResponse {
+    pub stacker_set: RewardSet,
+}
+
+impl GetStackersResponse {
+    pub fn load(
+        sortdb: &SortitionDB,
+        chainstate: &mut StacksChainState,
+        tip: &StacksBlockId,
+        burnchain: &Burnchain,
+        cycle_number: u64,
+    ) -> Result<Self, String> {
+        let cycle_start_height = burnchain.reward_cycle_to_block_height(cycle_number);
+
+        let pox_contract_name = burnchain
+            .pox_constants
+            .active_pox_contract(cycle_start_height);
+        let pox_version = PoxVersions::lookup_by_name(pox_contract_name)
+            .ok_or("Failed to lookup PoX contract version at tip")?;
+        if !matches!(pox_version, PoxVersions::Pox4) {
+            return Err(
+                "Active PoX contract version at tip is Pre-PoX-4, the signer set is not fetchable"
+                    .into(),
+            );
+        }
+
+        let provider = OnChainRewardSetProvider::new();
+        let stacker_set = provider.read_reward_set_nakamoto(
```

Comment:
This is something you could potentially cache in RAM in perpetuity.

Reply:
Sure -- but I'd prefer to leave that optimization for elsewhere.